File: mobile/android/app/src/main/java/com/xfshopee/CookieManager.kt
package com.xfshopee

import android.annotation.SuppressLint
import android.content.Context
import android.webkit.CookieManager as WebKitCookieManager
import android.webkit.ValueCallback

/**
 * Cookie管理器
 * 
 * 提供统一的Cookie管理功能，支持跨WebView的Cookie同步
 * 封装Android WebKit的CookieManager，提供更友好的API
 */
class CookieManager(private val context: Context) {
    
    companion object {
        // 默认域名
        private const val DEFAULT_DOMAIN = "a.b.c"
        private const val DEFAULT_PATH = "/"
    }
    
    // Android WebKit CookieManager
    private val webKitCookieManager: WebKitCookieManager by lazy {
        WebKitCookieManager.getInstance().apply {
            setAcceptCookie(true)
            setAcceptThirdPartyCookies(null, true)
        }
    }
    
    // 内存中的Cookie缓存（用于快速访问）
    private val cookieCache: MutableMap<String, MutableMap<String, String>> = mutableMapOf()
    
    /**
     * 设置Cookie
     * @param domain 域名（如 "a.b.c"）
     * @param cookieString Cookie字符串（如 "session_token=abc123; path=/; secure"）
     */
    fun setCookie(domain: String, cookieString: String) {
        val url = if (domain.startsWith("http")) domain else "https://$domain"
        
        // 设置到WebKit
        webKitCookieManager.setCookie(url, cookieString)
        
        // 更新缓存
        updateCache(domain, cookieString)
    }
    
    /**
     * 设置Cookie（简化版）
     */
    fun setCookie(name: String, value: String, domain: String = DEFAULT_DOMAIN) {
        val cookieString = "$name=$value; path=$DEFAULT_PATH; domain=$domain"
        setCookie(domain, cookieString)
    }
    
    /**
     * 获取指定域名的所有Cookie
     */
    fun getCookies(domain: String = DEFAULT_DOMAIN): List<String> {
        val url = if (domain.startsWith("http")) domain else "https://$domain"
        val cookies = webKitCookieManager.getCookie(url)
        
        return cookies?.split(";")?.map { it.trim() } ?: emptyList()
    }
    
    /**
     * 获取指定Cookie的值
     */
    fun getCookie(name: String, domain: String = DEFAULT_DOMAIN): String? {
        // 先从缓存查找
        val domainCookies = cookieCache[domain]
        if (domainCookies != null && domainCookies.containsKey(name)) {
            return domainCookies[name]
        }
        
        // 从WebKit获取
        val cookies = getCookies(domain)
        for (cookie in cookies) {
            if (cookie.startsWith("$name=")) {
                val value = cookie.substringAfter("=")
                
                // 更新缓存
                val cache = cookieCache.getOrPut(domain) { mutableMapOf() }
                cache[name] = value
                
                return value
            }
        }
        
        return null
    }
    
    /**
     * 移除指定Cookie
     */
    fun removeCookie(name: String, domain: String = DEFAULT_DOMAIN) {
        // 设置过期时间（过去的时间）来删除Cookie
        val expiredCookie = "$name=; path=$DEFAULT_PATH; domain=$domain; expires=Thu, 01 Jan 1970 00:00:00 GMT"
        setCookie(domain, expiredCookie)
        
        // 从缓存移除
        cookieCache[domain]?.remove(name)
    }
    
    /**
     * 清空指定域名的所有Cookie
     */
    fun clearCookies(domain: String = DEFAULT_DOMAIN) {
        val url = if (domain.startsWith("http")) domain else "https://$domain"
        
        // 获取所有Cookie并逐个删除
        val cookies = getCookies(domain)
        for (cookie in cookies) {
            val name = cookie.substringBefore("=").trim()
            removeCookie(name, domain)
        }
        
        // 清空缓存
        cookieCache.remove(domain)
    }
    
    /**
     * 清空所有Cookie
     */
    @SuppressLint("NewApi")
    fun clearAllCookies() {
        // 清空WebKit Cookie
        webKitCookieManager.removeAllCookies(null)
        webKitCookieManager.flush()
        
        // 清空缓存
        cookieCache.clear()
    }
    
    /**
     * 将Cookie注入到所有WebView
     */
    fun syncToAllWebViews(webViews: List<android.webkit.WebView>) {
        // WebKit CookieManager已经全局管理Cookie
        // 这里只需要确保所有WebView都启用了Cookie
        webViews.forEach { webView ->
            webView.settings.apply {
                javaScriptEnabled = true
                domStorageEnabled = true
            }
        }
        
        // 刷新Cookie
        webKitCookieManager.flush()
    }
    
    /**
     * 从WebView同步Cookie到管理器
     */
    fun syncFromWebView(webView: android.webkit.WebView, domain: String = DEFAULT_DOMAIN) {
        // WebKit CookieManager是全局的，无需手动同步
        // 但可以更新缓存
        val url = webView.url ?: return
        if (url.contains(domain)) {
            val cookies = webKitCookieManager.getCookie(url)
            if (cookies != null) {
                updateCache(domain, cookies)
            }
        }
    }
    
    /**
     * 导出所有Cookie（用于调试或备份）
     */
    fun exportAllCookies(): Map<String, List<String>> {
        val result = mutableMapOf<String, List<String>>()
        
        // 导出缓存中的所有域名
        cookieCache.keys.forEach { domain ->
            result[domain] = getCookies(domain)
        }
        
        return result
    }
    
    /**
     * 导入Cookie（用于恢复）
     */
    fun importCookies(cookies: Map<String, List<String>>) {
        cookies.forEach { (domain, cookieList) ->
            cookieList.forEach { cookieString ->
                setCookie(domain, cookieString)
            }
        }
    }
    
    /**
     * 检查Cookie是否已设置
     */
    fun hasCookie(name: String, domain: String = DEFAULT_DOMAIN): Boolean {
        return getCookie(name, domain) != null
    }
    
    /**
     * 获取Cookie的过期时间（如果设置）
     */
    fun getCookieExpiry(name: String, domain: String = DEFAULT_DOMAIN): Long? {
        // 简化实现：返回null表示会话Cookie
        // 在实际应用中，需要解析Cookie字符串中的expires/max-age属性
        return null
    }
    
    /**
     * 设置安全Cookie（HTTPS only）
     */
    fun setSecureCookie(name: String, value: String, domain: String = DEFAULT_DOMAIN) {
        val cookieString = "$name=$value; path=$DEFAULT_PATH; domain=$domain; secure"
        setCookie(domain, cookieString)
    }
    
    /**
     * 设置HttpOnly Cookie
     * 注意：Android WebKit不支持通过JavaScript设置HttpOnly标志
     * 这个标志通常在服务器设置
     */
    fun setHttpOnlyCookie(name: String, value: String, domain: String = DEFAULT_DOMAIN) {
        val cookieString = "$name=$value; path=$DEFAULT_PATH; domain=$domain; httponly"
        setCookie(domain, cookieString)
    }
    
    /**
     * 设置带过期时间的Cookie
     */
    fun setCookieWithExpiry(
        name: String,
        value: String,
        expirySeconds: Long,
        domain: String = DEFAULT_DOMAIN
    ) {
        val expiryDate = java.util.Date(System.currentTimeMillis() + expirySeconds * 1000)
        val expiryString = java.text.SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss z", java.util.Locale.US).format(expiryDate)
        
        val cookieString = "$name=$value; path=$DEFAULT_PATH; domain=$domain; expires=$expiryString"
        setCookie(domain, cookieString)
    }
    
    /**
     * 更新缓存
     */
    private fun updateCache(domain: String, cookieString: String) {
        val cookies = cookieString.split(";")
        val cache = cookieCache.getOrPut(domain) { mutableMapOf() }
        
        cookies.forEach { cookie ->
            if (cookie.contains("=")) {
                val parts = cookie.split("=", limit = 2)
                if (parts.size == 2) {
                    val name = parts[0].trim()
                    val value = parts[1].trim()
                    
                    // 检查是否是删除操作（值为空）
                    if (value.isEmpty() || cookie.contains("expires=Thu, 01 Jan 1970")) {
                        cache.remove(name)
                    } else {
                        cache[name] = value
                    }
                }
            }
        }
    }
    
    /**
     * 打印所有Cookie（调试用）
     */
    fun printAllCookies() {
        println("=== All Cookies ===")
        cookieCache.forEach { (domain, cookies) ->
            println("Domain: $domain")
            cookies.forEach { (name, value) ->
                println("  $name=$value")
            }
        }
        println("===================")
    }
    
    /**
     * 清除过期Cookie
     */
    fun clearExpiredCookies() {
        // 简化实现：清除所有Cookie中过期的
        // 在实际应用中，需要检查每个Cookie的expires属性
        val now = System.currentTimeMillis()
        
        cookieCache.forEach { (domain, cookies) ->
            // 这里可以添加逻辑检查Cookie是否过期
            // 暂时不清除，因为Android WebKit会自动管理
        }
    }
}

File: mobile/ios/Native/CookieManager.swift
import UIKit
import WebKit

/**
 * Cookie管理器
 * 
 * 提供统一的Cookie管理功能，支持跨WebView的Cookie同步
 */
class CookieManager: NSObject {
    
    // 单例实例
    static let shared = CookieManager()
    
    // HTTP Cookie存储
    private let httpCookieStorage = HTTPCookieStorage.shared
    
    // 默认域名
    private let defaultDomain = "a.b.c"
    
    private override init() {
        super.init()
        setupCookiePolicy()
    }
    
    /**
     * 设置Cookie策略
     */
    private func setupCookiePolicy() {
        // 允许所有Cookie
        httpCookieStorage.cookieAcceptPolicy = .always
    }
    
    /**
     * 设置Cookie
     */
    func setCookie(_ cookieString: String, for domain: String? = nil) {
        let targetDomain = domain ?? defaultDomain
        
        // 解析Cookie字符串
        let cookieComponents = cookieString.split(separator: ";").map { $0.trimmingCharacters(in: .whitespaces) }
        
        guard !cookieComponents.isEmpty else {
            return
        }
        
        // 第一个组件是name=value
        let nameValue = cookieComponents[0].split(separator: "=", maxSplits: 1)
        guard nameValue.count == 2 else {
            return
        }
        
        let name = String(nameValue[0])
        let value = String(nameValue[1])
        
        // 构建Cookie属性字典
        var properties: [HTTPCookiePropertyKey: Any] = [
            .name: name,
            .value: value,
            .domain: targetDomain,
            .path: "/"
        ]
        
        // 解析其他属性
        for component in cookieComponents.dropFirst() {
            if component.lowercased().hasPrefix("domain=") {
                let domainValue = String(component.dropFirst(7))
                if !domainValue.isEmpty {
                    properties[.domain] = domainValue
                }
            } else if component.lowercased().hasPrefix("path=") {
                let pathValue = String(component.dropFirst(5))
                if !pathValue.isEmpty {
                    properties[.path] = pathValue
                }
            } else if component.lowercased().hasPrefix("expires=") {
                let expiresValue = String(component.dropFirst(8))
                if let date = parseDate(expiresValue) {
                    properties[.expires] = date
                }
            } else if component.lowercased().hasPrefix("max-age=") {
                let maxAgeValue = String(component.dropFirst(8))
                if let maxAge = Int(maxAgeValue) {
                    let expiresDate = Date().addingTimeInterval(TimeInterval(maxAge))
                    properties[.expires] = expiresDate
                }
            } else if component.lowercased() == "secure" {
                properties[.secure] = true
            } else if component.lowercased() == "httponly" {
                properties[.httpOnly] = true
            }
        }
        
        // 创建并存储Cookie
        if let cookie = HTTPCookie(properties: properties) {
            httpCookieStorage.setCookie(cookie)
            synchronizeToWebKit()
        }
    }
    
    /**
     * 设置Cookie (简化版)
     */
    func setCookie(name: String, value: String, domain: String? = nil, path: String = "/") {
        let targetDomain = domain ?? defaultDomain
        
        let properties: [HTTPCookiePropertyKey: Any] = [
            .name: name,
            .value: value,
            .domain: targetDomain,
            .path: path
        ]
        
        if let cookie = HTTPCookie(properties: properties) {
            httpCookieStorage.setCookie(cookie)
            synchronizeToWebKit()
        }
    }
    
    /**
     * 获取指定域名的所有Cookie
     */
    func getCookies(for domain: String? = nil) -> [String] {
        let targetDomain = domain ?? defaultDomain
        let cookies = httpCookieStorage.cookies ?? []
        
        return cookies.filter { $0.domain.contains(targetDomain) }
            .map { "\($0.name)=\($0.value)" }
    }
    
    /**
     * 获取指定Cookie的值
     */
    func getCookie(named name: String, domain: String? = nil) -> String? {
        let targetDomain = domain ?? defaultDomain
        let cookies = httpCookieStorage.cookies ?? []
        
        return cookies.first { $0.name == name && $0.domain.contains(targetDomain) }?.value
    }
    
    /**
     * 移除指定Cookie
     */
    func removeCookie(named name: String, domain: String? = nil) {
        let targetDomain = domain ?? defaultDomain
        let cookies = httpCookieStorage.cookies ?? []
        
        if let cookie = cookies.first(where: { $0.name == name && $0.domain.contains(targetDomain) }) {
            httpCookieStorage.deleteCookie(cookie)
            synchronizeToWebKit()
        }
    }
    
    /**
     * 清空指定域名的所有Cookie
     */
    func clearCookies(for domain: String? = nil) {
        let targetDomain = domain ?? defaultDomain
        let cookies = httpCookieStorage.cookies ?? []
        
        for cookie in cookies where cookie.domain.contains(targetDomain) {
            httpCookieStorage.deleteCookie(cookie)
        }
        
        synchronizeToWebKit()
    }
    
    /**
     * 清空所有Cookie
     */
    func clearAllCookies() {
        let cookies = httpCookieStorage.cookies ?? []
        
        for cookie in cookies {
            httpCookieStorage.deleteCookie(cookie)
        }
        
        synchronizeToWebKit()
        
        // 同时清除WKWebsiteDataStore中的Cookie
        clearWebKitCookies()
    }
    
    /**
     * 将Cookie同步到WebKit (WKWebView)
     */
    func synchronizeToWebKit() {
        // HTTP Cookie存储会自动同步到WebKit
        // 但我们可以强制同步
        httpCookieStorage.cookies?.forEach { cookie in
            WKWebsiteDataStore.default().httpCookieStore.setCookie(cookie)
        }
    }
    
    /**
     * 清除WebKit中的Cookie
     */
    private func clearWebKitCookies() {
        let dataStore = WKWebsiteDataStore.default()
        
        dataStore.fetchDataRecords(ofTypes: WKWebsiteDataStore.allWebsiteDataTypes()) { records in
            dataStore.removeData(ofTypes: WKWebsiteDataStore.allWebsiteDataTypes(),
                                for: records) {
                print("WebKit cookies cleared")
            }
        }
    }
    
    /**
     * 注入Cookie到WebView
     */
    func injectCookies(to webView: WKWebView, domain: String? = nil) {
        let targetDomain = domain ?? defaultDomain
        let cookies = httpCookieStorage.cookies ?? []
        
        for cookie in cookies where cookie.domain.contains(targetDomain) {
            webView.configuration.websiteDataStore.httpCookieStore.setCookie(cookie)
        }
    }
    
    /**
     * 从WebView同步Cookie
     */
    func syncFromWebView(_ webView: WKWebView, domain: String? = nil) {
        let targetDomain = domain ?? defaultDomain
        let cookieStore = webView.configuration.websiteDataStore.httpCookieStore
        
        cookieStore.getAllCookies { cookies in
            for cookie in cookies where cookie.domain.contains(targetDomain) {
                self.httpCookieStorage.setCookie(cookie)
            }
        }
    }
    
    /**
     * 导出所有Cookie (用于调试或备份)
     */
    func exportAllCookies() -> [String: [String]] {
        let cookies = httpCookieStorage.cookies ?? []
        var result: [String: [String]] = [:]
        
        for cookie in cookies {
            var cookieStrings = result[cookie.domain] ?? []
            cookieStrings.append("\(cookie.name)=\(cookie.value)")
            result[cookie.domain] = cookieStrings
        }
        
        return result
    }
    
    /**
     * 导入Cookie (用于恢复)
     */
    func importCookies(_ cookies: [String: [String]]) {
        for (domain, cookieStrings) in cookies {
            for cookieString in cookieStrings {
                setCookie(cookieString, for: domain)
            }
        }
    }
    
    /**
     * 检查Cookie是否已设置
     */
    func hasCookie(named name: String, domain: String? = nil) -> Bool {
        return getCookie(named: name, domain: domain) != nil
    }
    
    /**
     * 获取Cookie的过期时间
     */
    func getCookieExpiry(named name: String, domain: String? = nil) -> Date? {
        let targetDomain = domain ?? defaultDomain
        let cookies = httpCookieStorage.cookies ?? []
        
        return cookies.first { $0.name == name && $0.domain.contains(targetDomain) }?.expiresDate
    }
    
    /**
     * 设置安全Cookie (HTTPS only)
     */
    func setSecureCookie(name: String, value: String, domain: String? = nil) {
        let targetDomain = domain ?? defaultDomain
        
        let properties: [HTTPCookiePropertyKey: Any] = [
            .name: name,
            .value: value,
            .domain: targetDomain,
            .path: "/",
            .secure: true
        ]
        
        if let cookie = HTTPCookie(properties: properties) {
            httpCookieStorage.setCookie(cookie)
            synchronizeToWebKit()
        }
    }
    
    /**
     * 设置HttpOnly Cookie
     */
    func setHttpOnlyCookie(name: String, value: String, domain: String? = nil) {
        let targetDomain = domain ?? defaultDomain
        
        let properties: [HTTPCookiePropertyKey: Any] = [
            .name: name,
            .value: value,
            .domain: targetDomain,
            .path: "/",
            .httpOnly: true
        ]
        
        if let cookie = HTTPCookie(properties: properties) {
            httpCookieStorage.setCookie(cookie)
            synchronizeToWebKit()
        }
    }
    
    /**
     * 设置带过期时间的Cookie
     */
    func setCookieWithExpiry(name: String, value: String, expirySeconds: TimeInterval, domain: String? = nil) {
        let targetDomain = domain ?? defaultDomain
        let expiryDate = Date().addingTimeInterval(expirySeconds)
        
        let properties: [HTTPCookiePropertyKey: Any] = [
            .name: name,
            .value: value,
            .domain: targetDomain,
            .path: "/",
            .expires: expiryDate
        ]
        
        if let cookie = HTTPCookie(properties: properties) {
            httpCookieStorage.setCookie(cookie)
            synchronizeToWebKit()
        }
    }
    
    /**
     * 解析日期字符串
     */
    private func parseDate(_ dateString: String) -> Date? {
        let formatters = [
            "EEE, dd MMM yyyy HH:mm:ss zzz",
            "EEEE, dd-MMM-yy HH:mm:ss zzz",
            "EEE MMM dd HH:mm:ss yyyy"
        ]
        
        for format in formatters {
            let formatter = DateFormatter()
            formatter.dateFormat = format
            formatter.locale = Locale(identifier: "en_US_POSIX")
            formatter.timeZone = TimeZone(secondsFromGMT: 0)
            
            if let date = formatter.date(from: dateString) {
                return date
            }
        }
        
        return nil
    }
    
    /**
     * 打印所有Cookie (调试用)
     */
    func printAllCookies() {
        let cookies = httpCookieStorage.cookies ?? []
        
        print("=== All Cookies (\(cookies.count) items) ===")
        for cookie in cookies {
            print("Domain: \(cookie.domain)")
            print("  Name: \(cookie.name)")
            print("  Value: \(cookie.value)")
            print("  Path: \(cookie.path)")
            print("  Secure: \(cookie.isSecure)")
            print("  HTTPOnly: \(cookie.isHTTPOnly)")
            if let expires = cookie.expiresDate {
                print("  Expires: \(expires)")
            }
            print("---")
        }
        print("========================================")
    }
}